packages/tpetra/core/src/Tpetra_Details_scaleBlockDiagonal.hpp

```diff
@@ -0,0 +1,111 @@
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_SCALEBLOCKDIAGONAL_HPP
+#define TPETRA_DETAILS_SCALEBLOCKDIAGONAL_HPP
+
+#include "TpetraCore_config.h"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Teuchos_ScalarTraits.hpp"
+#include "Tpetra_Details_Behavior.hpp"
+#include "KokkosBatched_Util.hpp"
+#include "KokkosBatched_LU_Decl.hpp"
+#include "KokkosBatched_LU_Serial_Impl.hpp"
+#include "KokkosBatched_Trsm_Decl.hpp"
+#include "KokkosBatched_Trsm_Serial_Impl.hpp"
+
+
+/// \file Tpetra_Details_scaleBlockDiagonal.hpp
+/// \brief Functions that rescales a multivector (in-place) by a inverse of a block-diagonal matrix 
+/// (as implied by the given MultiVector
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra.  The file itself or its contents may disappear or
+///   change at any time.
+
+namespace Tpetra {
+namespace Details {
+
+
+template<class MultiVectorType>        
+void inverseScaleBlockDiagonal(const MultiVectorType & blockDiagonal, MultiVectorType & multiVectorToBeScaled) {
+  using SC             = typename MultiVectorType::scalar_type;
+  using LO             = typename MultiVectorType::local_ordinal_type;
+  using local_mv_type  = typename MultiVectorType::dual_view_type::t_dev;
+  using range_type     = Kokkos::RangePolicy<typename MultiVectorType::node_type::execution_space, LO>;
+  using namespace KokkosBatched;
+  SC SC_one = Teuchos::ScalarTraits<SC>::one();
+
+  // Sanity checking: Map Compatibility (A's rowmap matches diagonal's map)
+  if (Tpetra::Details::Behavior::debug() == true) {
+    TEUCHOS_TEST_FOR_EXCEPTION(!blockDiagonal.getMap()->isSameAs(*multiVectorToBeScaled.getMap()),
+       std::runtime_error, "Tpetra::Details::scaledBlockDiagonal was given incompatible maps");
+  }
+
+  LO numrows   = blockDiagonal.getLocalLength();
+  LO blocksize = blockDiagonal.getNumVectors();
+  LO numblocks = numrows / blocksize;
+
+  // Get Kokkos versions of objects
+  local_mv_type blockDiag = blockDiagonal.getLocalViewDevice();
+  local_mv_type toScale   = multiVectorToBeScaled.getLocalViewDevice();
+  
+  typedef Algo::Level3::Unblocked algo_type;
+  Kokkos::parallel_for("scaleBlockDiagonal",range_type(0,numblocks),KOKKOS_LAMBDA(const LO i){
+      std::pair<LO,LO> row_range = std::make_pair(i*blocksize,(i+1)*blocksize);
+      auto A = Kokkos::subview(blockDiag,row_range, Kokkos::ALL());
+      auto B = Kokkos::subview(toScale,  row_range, Kokkos::ALL());
+
+      // Factor
+      SerialLU<algo_type>::invoke(A);
+
+      // Solve L
+      SerialTrsm<Side::Left,Uplo::Lower,Trans::NoTranspose,Diag::Unit,algo_type>::invoke(SC_one,A,B);
+
+      // Solve U
+      SerialTrsm<Side::Left,Uplo::Upper,Trans::NoTranspose,Diag::NonUnit,algo_type>::invoke(SC_one,A,B);
```

Comment:
`SC_one` has type `std::complex<T>`, not `Kokkos::complex<T>`, so it's not OK to use in a device kernel.  `impl_scalar_type` works fine here.

Reply:
Fixed.